packages/contracts/contracts/AppInstanceAdjudicator.sol

```diff
@@ -0,0 +1,582 @@
+  /// This value must have been signed off by all parties to the channel, that is,
+  /// this must be called with the correct msg.sender or signatures must be provided.
+  /// @param appStateHash The hash of the agreed upon state. Typically this is the latest signed state.
+  /// @param nonce The nonce of the agreed upon state
+  /// @param timeout A dynamic timeout value representing the timeout for this state
+  /// @param signatures A sorted bytes string of concatenated signatures of each signingKey
+  /// @dev This function is only callable when the state channel is in an ON state.
+  function setState(
+    bytes32 _id,
+    bytes32 appStateHash,
+    uint256 nonce,
+    uint256 timeout,
+    bytes signatures
+  )
+    public
+    onlyWhenChannelOpen(_id)
+  {
+    AppInstance storage c = apps[_id];
+
+    if (msg.sender != c.auth.owner) {
+      bytes32 h = computeStateHash(_id, appStateHash, nonce, timeout);
+      require(
+        signatures.verifySignatures(h, c.auth.signingKeys),
+        "Invalid signatures"
+      );
+    }
+
+    if (timeout > 0) {
+      require(
+        nonce > c.state.nonce,
+        "Tried to set state with non-new state"
+      );
+      c.state.status = Status.DISPUTE;
+    } else {
+      require(
+        nonce >= c.state.nonce,
+        "Tried to finalize state with stale state"
+      );
+      c.state.status = Status.OFF;
+    }
+
+    c.state.appStateHash = appStateHash;
+    c.state.nonce = nonce;
+    c.state.disputeNonce = 0;
+    c.state.finalizesAt = block.number + timeout;
+    c.state.disputeCounter += 1;
+    c.state.latestSubmitter = msg.sender;
+  }
+
+  /// @notice Create a dispute regarding the latest signed state and immediately after,
+  /// performs a unilateral action to update it.
+  /// @param app An `App` struct specifying the application logic
+  /// @param appState The ABI encoded application state
+  /// @param nonce The nonce of the agreed upon state
+  /// @param action The ABI encoded action the submitter wishes to take
+  /// @param appStateSignatures A sorted bytes string of concatenated signatures on the
+  /// `appState` state, signed by all `signingKeys`
+  /// @param actionSignature A bytes string of a single signature by the address of the
+  /// signing key for which it is their turn to take the submitted `action`
+  /// @param claimFinal A boolean representing a claim by the caller that the action
+  /// progresses the state of the application to a terminal / finalized state
+  /// @dev Note this function is only callable when the state channel is in an ON state
+  function createDispute(
+    bytes32 _id,
+    App app,
+    bytes appState,
+    uint256 nonce,
+    bytes action,
+    bytes appStateSignatures,
+    bytes actionSignature,
+    bool claimFinal
+  )
+    public
+    onlyWhenChannelOpen(_id)
+  {
+
+    // FIXME: Before merging, the timeout variable has been removed to get
+    // a stack too deep error to go away ... need to refactor function.
```

Comment:
Is solving this error just a matter of adding an extra function?

Reply:
That would be an unfortunate way of solving it. Seems like I can place more local variable inside structs to solve it.